src/Ouroboros/Network/Socket.hs

```diff
@@ -0,0 +1,469 @@
+type PongSideProtocol = Protocol MsgPong MsgPing
+type PingSideProtocol = Protocol MsgPing MsgPong
+
+data MsgPong = MsgPong deriving (Eq, Show)
+data MsgPing = MsgPing deriving (Eq, Show)
+
+instance Serialise MsgPong where
+  encode _ = encodeNull
+  decode = MsgPong <$ decodeNull
+
+instance Serialise MsgPing where
+  encode _ = encodeNull
+  decode = MsgPing <$ decodeNull
+
+
+
+runProducer :: forall block r. (Chain.HasHeader block, Serialise block)
+         => Socket -> ProducerHandlers block IO r -> IO [Async ()]
+runProducer sd producer = do
+    sem <- newQSem 0
+    (chain_tid, chain_wq, chain_rq) <- startupConversation sem ChainHeaderSyncProducer
+                                       producerSideProtocol
+    (pong_tid, pong_wq, pong_rq) <- startupConversation sem Ponger pong
+    reader_tid <- startupReader sd $ M.fromList [ (ChainHeaderSyncConsumer, chain_rq)
+                                                , (Pinger, pong_rq)]
+    wtid <- startupWriter sd sem [(ChainHeaderSyncProducer, chain_wq), (Ponger, pong_wq)]
+    return [wtid, chain_tid, pong_tid, reader_tid]
+
+  where
+    -- Reuse the generic 'producerSideProtocol1'
+    -- but interpret it in our Protocol free monad.
+    producerSideProtocol :: ProducerSideProtocol block ()
+    producerSideProtocol =
+      producerSideProtocol1
+        (liftProducerHandlers liftIO producer)
+        sendMsg
+        recvMsg
+
+runConsumer :: forall block. (Chain.HasHeader block, Serialise block)
+         => Socket -> ConsumerHandlers block IO -> IO [Async ()]
+runConsumer sd consumer = do
+    sem <- newQSem 0
+    (chain_tid, chain_wq, chain_rq) <- startupConversation sem ChainHeaderSyncConsumer
+                                       consumerSideProtocol
+    (ping_tid, ping_wq, ping_rq) <- startupConversation sem Pinger ping
+    reader_tid <- startupReader sd $ M.fromList [ (ChainHeaderSyncProducer, chain_rq)
+                                                , (Ponger, ping_rq)]
+    wtid <- startupWriter sd sem [(ChainHeaderSyncConsumer, chain_wq), (Pinger, ping_wq)]
+
+    return [wtid, chain_tid, ping_tid, reader_tid]
+
+  where
+    -- Reuse the generic 'consumerSideProtocol1'
+    -- but interpret it in our Protocol free monad.
+    consumerSideProtocol :: ConsumerSideProtocol block ()
+    consumerSideProtocol =
+      consumerSideProtocol1
+        (liftConsumerHandlers liftIO consumer)
+        sendMsg
+        recvMsg
+
+------------------------------------------------
+
+
+runProtocolWithTBQueues :: forall smsg rmsg.  (Serialise smsg, Serialise rmsg)
+                        => QSem
+                        -> (BS.ByteString -> STM ())
+                        -> STM BS.ByteString
+                        -> Protocol smsg rmsg ()
+                        -> IO ()
+runProtocolWithTBQueues sem wqueue rqueue p =
+    unProtocol p >>= go mempty
+  where
+    go trailing (Send msg k) = do
+      let body = BL.toStrict $ CBOR.toLazyByteString (encode msg)
```

Comment:
`BL.toStrict` is expensive, is it necessary?

Reply:
I'm not sure that it is.